service/security/shiro/src/main/java/org/eclipse/kapua/service/authentication/shiro/realm/UserPassCredentialsMatcher.java

```diff
@@ -179,3 +199,42 @@ public boolean doCredentialsMatch(AuthenticationToken authenticationToken, Authe
     }
 
 }
+
+interface PasswordChecker {
+
+    boolean checkPassword(String tokenPassword, Credential infoCredential);
+
+}
+
+class CachePasswordChecker implements PasswordChecker {
+
+    private static final Cache<String, String> CACHED_CREDENTIALS = new LocalCache<String, String>(
+            KapuaAuthenticationSetting.getInstance().getInt(KapuaAuthenticationSettingKeys.AUTHENTICATION_CREDENTIAL_USERPASS_CACHE_CACHE_SIZE, 1000),
+            KapuaAuthenticationSetting.getInstance().getInt(KapuaAuthenticationSettingKeys.AUTHENTICATION_CREDENTIAL_USERPASS_CACHE_CACHE_TTL, 60),
+            null);
+
+    public boolean checkPassword(String tokenPassword, Credential infoCredential) {
+        String hashedTokenPassword = Base64.getEncoder().encodeToString(DigestUtils.sha3_512(tokenPassword));
+        String cachedCredential = CACHED_CREDENTIALS.get(hashedTokenPassword);
+        if (cachedCredential!=null && cachedCredential.equals(infoCredential.getCredentialKey())) {
+            return true;
+        }
+        else if (BCrypt.checkpw(tokenPassword, infoCredential.getCredentialKey())) {
+            CACHED_CREDENTIALS.put(hashedTokenPassword, infoCredential.getCredentialKey());
```

Comment:
It looks like we are using the cache when the correct password is used.  However, consider the case where devices have been configured with the wrong password (or it has been changed, has expired, etc.).  I think we should also be using the cache for when the wrong credentials are provided.

Reply:
In other words, you'd cached the result of the match, is that right? Either way, we need to be wary of the cache refresh strategy of the cache.